### ISSUE TITLE
Support non-merging flows

### ISSUE THREAD

**Issue by hjoliver** [cylc/cylc-flow#4419]
Currently, `cylc trigger --reflow` can trigger multiple flows within the same graph, and if one flow catches up to another they merge together (one task at a time: e.g. if `foo.10` gets spawned in flow `m` and we find that `foo.10` already exists in the active task pool and belongs to flow `n`, we end up with a single `foo.10` instance that belongs to flows `{m,n}`).

The main use cases for reflow initially are likely to be:
- rewind the workflow without checkpoint restart or warm start from cycle point, in which case the original flow will be stopped
- re-run a dead-end side branch of graph, e.g. to regenerate products for one cycle point after changing something on disk

Neither of these require flow merging as a feature in itself. Merging is really just a way to prevent anything pathological happening in the unlikely event that users do trigger flows that end up in the same part of the graph.

However, as an alternative to merging we could (should IMO!) allow non-interacting flows that can pass through each other without merging or interfering.  Technically this would be easy to do (easier than merging) in the scheduler - it just requires making flow number part of the formal task ID so that (e.g.) `foo.10{m}` and `foo.10{n}` can co-exist in the task pool. 

This would allow a new kind of cycling. For example, a simple finite graph could process multiple datasets at once using a different flow for each one. 

How to represent this in the UI might be tricky.

**Comment by hjoliver** [comment]
Note if we do this (and put flow number in log path etc. too), job number could increment from `01` again for each new flow. (Not strictly necessary though).

**Comment by hjoliver** [comment]
If we do this, it might make sense to switch entirely to non-merging flows rather support both merging and non-merging flows as options.

Note I expected that users would generally want to avoid one flow from catching up to another, and implemented flow merging to prevent anything pathological going down if it does happen. Non-merging flows probably make more sense, but I didn't go there at the time because of complicated task ID and job log path implications: 

*For non-merging flows, flow number would logically be part of the universal task/job ID: `flow/cycle/task/job`.*  Internally at least, the unique task ID has to include flow number in order to allow multiple instances (from different flows) of the same `task.point` to co-exist in the scheduler task pool.

So:
- if we stick with merging flows only, the flow number does not need to be part of the task/job ID
- if we intend to replace merging flows with non-merging ones, it would make sense to put flow number in the ID
- if we intend to support both merging and non-merging flows at once, the ID situation is less clear as merged sets of flow numbers make for an ugly ID :vomiting_face: 

Note we can do any of the above without including flow number in the ID, but for the non-merging case it would be sensible and simpler to do it.  And as argued above, non-merging flows probably are what we should be aiming for.

**Comment by oliver-sanders** [comment]
Agreed with all the above, except that I think merging flows are quite useful. E.G. one of my tasks produced bad output for some reason, so I fix the problem and rerun everything downstream of that task in a reflow until the new flow catches up with the old one.

**Comment by hjoliver** [comment]
> We would also need to add the flow into the log/job, work/ and probably share directories too.

Yeah that's what I was getting at with this comment above: *... task ID and job log path implications*

> I think merging flows are quite useful. E.G. one of my tasks produced bad output for some reason, so I fix the problem and rerun everything downstream of that task in a reflow until the new flow catches up with the old one.

This could be done with non-merging flows too, although it would require more careful planning or intervention (to get the reflow to stop at the right place). But yes, merging seems the simplest way to achieve this.

**Comment by hjoliver** [comment]
Paste from @oliver-sanders comments on Element chat:

----

Yeah, I had considered this.

Reflow developed alongside the UID proposal so the UID does not contain it (& the existing ID systems don't either).

How to fit the flow in with the universal ID is tricky. The easiest solution is to make it orthogonal and use --flow, but if we want non-merging flows then we have the potential for duplicate tasks in different flows which, at present would break the task pool, data store, job submission logic and much more.

Fitting the flow into the UID is a challenge but a solvable one, some ideas below. However, the UID is not the biggest problem here. To support this properly we would need to add another level to the log/job and work (and share?) hierarchies (otherwise the duplicate tasks would be fighting over the same files) and make changes to the DB, etc.

So I think this is going to have to be a **Cylc 9** problem as there is no way we can make these changes in time for Cylc 8 and the changes will definitely be breaking. However, I can make sure to leave room in the UID work now to open the door to this in the future.

The full form of the universal ID is:

```
~user/workflow//cycle/task/job
```
The flow would have to come after the `//` divider (because it needs to be in the relative part of the ID).

The logical solution would be to add an extra level:

```
//flow/cycle/task/job
```
However, that extra `1/` will make things much harder to understand at a glance. task.cycle and cycle/task are both pretty easy to scan, but beyond that it gets hard to pull out the important information. Multiple flows are the exception not the rule so making things confusing for the most likely cases to cater to the least likely is undesirable.

We could use a sort of modifier to patch it into the existing hierarchy like so:

```
//cycle^flow/task/job
```

Although there is some ambiguity as to whether the modifier should go on the task or cycle. Perhaps easier to put it on the cycle internally and let it be set on either by users on the CLI.

Here are some CLI examples with the modifier approach above:

```
12345678# pause a task in flow "1"
cylc pause my_workflow//cycle/task

# pause a task in flow "2"
cylc pause my_workflow//cycle^2/task

# pause all tasks in flow "2"
cylc pause 'my_workflow//*^1/*'
```
We could also do it the other way around:
```
//[flow^]cycle/task/job
```

e.g.
```
123456# task in flow "1"
//cycle/task  # the 1^ is implicit
//1^cycle/task

# task in flow "2"
//2^cycle/task
```
CLI examples:
```
123456789101112# pause a task in flow "1"
cylc pause my_workflow//cycle/task   # the 1^ is implicit
cylc pause my_workflow//1^cycle/task

# pause a task in flow "2"
cylc pause my_workflow//2^cycle/task

# pause all tasks in flow "2"
cylc pause 'my_workflow//2^*/*'

# stop flow 2
cylc stop 'my_workflow//2^'
```

**Comment by hjoliver** [comment]
> >  I think merging flows are quite useful. E.G. one of my tasks produced bad output for some reason, so I fix the problem and rerun everything downstream of that task in a reflow until the new flow catches up with the old one.
> 
> This could be done with non-merging flows too, although it would require more careful planning or intervention (to get the reflow to stop at the right place). But yes, merging seems the simplest way to achieve this.

Probably a moot point till Cylc 9 anyway, but I think even this use case might be better done with non-merging flows.

At the moment (merging flows) you have to:
  1. pause the main flow, then let the new flow catch up and merge, before releasing it
  2. **or** let the main flow and new flow continue at the same time, and eventually merge

In case 1., you're unnecessarily delaying the new products just to regenerate the old ones 

In case 2., you'll likely repeat-run more tasks than is strictly necessary just to generate the old products (possibly mitigated by the wait on new real time data in real time workflows)

Re-generation of earlier products is almost certainly going to be cycle-point-specific, so the simplest way to handle it is: trigger the regeneration flow and tell it to stop at a certain cycle point (probably run just a single point in fact). 

The only time merging may be better, I think, is if the original flow has not got "out of the way" of the new flow yet, inside the regeneration window.
